Add MPIBlockDiag and MPILinearOperator

rohanbabbar04 opened the pull request:
- Added MPIBlockDiag with _matvec and _rmatvec which now uses `DistributedArray`
- Added MPILinearOperator class 
- Added basic classes(_ScaledLinearOperator, _AdjointLinearOperator, etc) and made them input and output `DistributedArray`.
- Added tests to verify above classes.
- Upgrade scipy to `1.8` and python to `3.8`
- Updated docs for sphinx

mrava87 commented on pylops_mpi/basicoperators/__init__.py:
```diff
@@ -0,0 +1 @@
+from .BlockDiag import MPIBlockDiag
```
I suggest to follow the structure of pylops https://github.com/PyLops/pylops/blob/dev/pylops/basicoperators/__init__.py

rohanbabbar04 replied:
Done...

mrava87 replied:
Great :)

mrava87 commented on pylops_mpi/__init__.py:
```diff
@@ -1,5 +1,6 @@
 from .DistributedArray import DistributedArray, Partition
 from .plotting.plotting import *
+from .basicoperators import MPIBlockDiag
```
Why not `from .basicoperators import *` so it is already the correct form for later

rohanbabbar04 replied:
Done...

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,115 @@
+
+from pylops_mpi.DistributedArray import local_split, Partition
+
+sp_version = sp.__version__.split(".")
```
Since we are building a new library, I suggest we do not support early versions of spicy (like we need to do for backward compatibility in pylops). Simply use `from scipy.sparse.linalg._interface import _get_dtype` and make 1.8.0 the minimum requirement

rohanbabbar04 replied:
Updated scipy to `1.8.0`, python is also now `3.8`(as required by scipy), updated the code, also updated `requirements-dev.txt`, `environment.yml` and `environment-dev.yml`. Also removed `3.7` from build.yaml.

mrava87 replied:
Perfect, I don't think we need to maintain python3.7 given that pretty much everyone uses 3.8+ these days...

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,115 @@
+
+        Notes
+        -----
+        Refer to :obj:pylops.BlockDiag for more details.
```
It is good to refer to pylops for the general idea of the operator, but you should add some content to explain how this specific operator works in a MPI fashion. Basically, how operators are splitted across workers, how the forward and adjoint work, etc. Try to spend more time on documentation, it is as IMPORTANT as code :)

rohanbabbar04 replied:
Updated the notes, would want your feedback...

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,115 @@
+        for iop, oper in enumerate(self.ops):
+            y.append(oper.matvec(x_local[self.mmops[iop]:
+                                         self.mmops[iop + 1]]))
+        if y:
```
I am not sure I understand this. In which scenario this if statement is not verified? (I guess if self.ops was empty?)

And if this is the case, what would happen to `np.hstack(self.base_comm.allgather(y))`

rohanbabbar04 replied:
Now it is removed  as we use DistributedArray.

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,115 @@
+
+    def _matvec(self, x):
+        # Calculate x_local using mops for different ranks
+        mops_sum = np.insert(np.cumsum(self.base_comm.allgather(self.mops)), 0, 0)
```
Seems like this can be done once upfront and saved into a class member?

rohanbabbar04 replied:
Removed as not required....

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,115 @@
+
+    def _rmatvec(self, x):
+        # Calculate x_local using nops for different ranks
+        nops_sum = np.insert(np.cumsum(self.base_comm.allgather(self.nops)), 0, 0)
```
Same as above

rohanbabbar04 replied:
Removed as not required

mrava87 reviewed (commented):
@rohanbabbar04, thanks for this PR. So far I struggle a bit to understand the rationale behind this operator. I had a few comments in the closed PR that I put here again, hopefully your answers (and example!) will help clarify my doubts.

- [Taken from closed PR] ALWAYS work with 1D vectors first! In your test if MPIBlockDiag you make a 3d array… when solving inverse problems we work with 1D arrays (internally operators may reshape them but from the outside - and from the solver point of view- they are 1D)… we need tests to prove that the operators we build work with 1D arrays (see tests in pylops where flatten/ravel is used in most places
- Looking at your example, I am not sure I understand how you envision/designed MPIBlockDiag… my idea is that if you have a block diagonal operator with N operators (N being the number of ranks), each rank will build one pylops operator and pass it to the MPIBlockDiag, so that once matvec/rmatvec is called each rank will compute their portion… so if you were to compare your new implementation with an equivalent pylops one, the pylops operator will have to take all of those operators and give the same result…. For simplicity take this example: you have 3 ranks and 3 5x5 matrices which you wrap into 3 MatrixMult operators. The first is filled with 1s, the second with 2s, the third with 3s.. now for pylops-MPI each rank makes one MatrixMult operator and passes it to MPIBlockDiag… the x is also distributed and can be just full of 1s for simplicity. Unless I don’t understand your example it doesn’t seem like it is built like this, rather each rank seems to make all of the ops…
- Based on the above, it seems like your `x` and `y` are not `DistributedArray`, instead just numpy arrays. Why? Again, I would think we should have distributed arrays passed to MPI operators, so that internally it is easy to get the portion needed
- In the PR you closed you had created a `LinearOperator` class and subclassed `MPIBlockDiag`. Was there any reason/benefit from doing that, or right now you do not see any use wrt `MPIBlockDiag`?

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,115 @@
+               Linear Operators allocated at each rank.
+        """
+        local_shapes = local_split((len(ops),), self.base_comm, Partition.SCATTER, 0)
+        x = np.insert(np.cumsum(self.base_comm.allgather(local_shapes)), 0, 0)
```
`x` is a bit of a unclear nomenclature, maybe you something a bit more verbose like `iops`

rohanbabbar04 replied:
Removed....

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        explicit : :obj:`bool`
+            Operator contains a matrix that can be solved explicitly (``True``) or not (``False``)
+
+        Notes
```
Very good :D

mrava87 replied:
There is also a slight inconsistency with the fact that we allow each rank to have not just one operator but a list of operators... let's maybe simply add.

```
Each rank has one or more :class:`pylops.LinearOperator`, which we represent here compactly as :math:`\mathbf{L}_i` for rank :math:`i`.

mrava87 replied:
Finally

**L** -> :math:`\mathbf{L}`

rohanbabbar04 replied:
Done :slightly_smiling_face:

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        self.nnops = np.insert(np.cumsum(nops), 0, 0)
+        self.mmops = np.insert(np.cumsum(mops), 0, 0)
+        # Shape of the operator is equal to the sum of nops and mops at all ranks
+        self.shape = (self.base_comm.allreduce(self.nops), self.base_comm.allreduce(self.mops))
```
you should not do self.shape, and then pass it to super... make just a temporary shape and then pass it. Same for dtype (basically the way you did clinear is how you should do all)

rohanbabbar04 replied:
Done....

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        super().__init__(shape=self.shape, dtype=self.dtype, clinear=clinear)
+
+    def _matvec(self, x: DistributedArray) -> DistributedArray:
+        if x.local_shape != (self.mops, ):
```
great!

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        for iop, oper in enumerate(self.ops):
+            y1.append(oper.matvec(x.local_array[self.mmops[iop]:
+                                                self.mmops[iop + 1]]))
+        if y1:
```
I still dont understand this... when y1 will be empty? and if it is empty, then what y would be (I guess just some zeros as you never assign anything to it?)

rohanbabbar04 replied:
Now since we are checking `x.local_shape != (self.mops, )` this should solve our problem 
Thus, removed `if y1`

mrava87 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        for iop, oper in enumerate(self.ops):
+            y1.append(oper.rmatvec(x.local_array[self.nnops[iop]:
+                                                 self.nnops[iop + 1]]))
+        if y1:
```
Same as above

rohanbabbar04 replied:
Similar to above, Since we are checking x.local_shape != (self.nops, ) this should solve our problem
 Thus, removed if y1

mrava87 commented on tests/test_blockdiag.py:
```diff
@@ -0,0 +1,49 @@
+        ops = [pylops.MatrixMult((i + 1) * np.ones(shape=(par['ny'], par['nx'])).astype(par['dtype'])) for i in range(size)]
+        BDiag = pylops.BlockDiag(ops=ops)
+
+        print(x_global.shape)
```
Remove this

mrava87 commented on tests/test_blockdiag.py:
```diff
@@ -0,0 +1,49 @@
+    y[:] = np.ones(shape=par['ny'], dtype=par['dtype'])
+    y_global = y.asarray()
+
+    x_mat = BDiag_MPI._matvec(x)
```
I did use `_matvec` in my quick tests because I realized we could not use ``matvec / @` without making a new LinearOperator, but ideally we should include it in this PR and then never use internal methods like `_matvec ` in tests :)

rohanbabbar04 replied:
Changed it from `_matvec` -> `@`
Changed it from `_rmatvec` -> `.H @`

rohanbabbar04 commented:
@mrava87 Done the changes, if the code and tests are good to go, we can merge this
I will make a new PR for the example(Post Stack one)

mrava87 reviewed (commented):
Good job! Still quite a few comments but I think we are getting closer to having a good MPIBlockDiag operator.

For now I think we can stick to this implementation. Later we can consider maybe adding also what you had originally, allowing users to choose (depending on the scenario) if they want to create&pass operators per rank or all of them from all ranks - the second may be more flexible for complex scenarios but I suspect the former will be the most used one as in most cases operators will be the same for all ranks, just filled with different 'content'. (or just doing the same think to different parts of the model and data vector)

rohanbabbar04 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+class MPIBlockDiag(LinearOperator):
+    r"""MPI Block-diagonal operator.
+
+        Create a block-diagonal operator from N linear operators using MPI.
```
Updated the docs...

rohanbabbar04 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        Parameters
+        ----------
+        ops : :obj:`list`
+            Linear operators to be stacked. Alternatively,
```
Changed to  One or more `class:pylops.LinearOperator`

rohanbabbar04 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+        ----------
+        ops : :obj:`list`
+            Linear operators to be stacked. Alternatively,
+            :obj:`numpy.ndarray` or :obj:`scipy.sparse` matrices can be passed
```
True, have changed that. In my earlier version i had that, forgot to remove it...
Done :)

rohanbabbar04 commented on pylops_mpi/basicoperators/BlockDiag.py:
```diff
@@ -0,0 +1,139 @@
+from pylops_mpi import DistributedArray
+
+
+class MPIBlockDiag(LinearOperator):
```
Changed Now a New class is created `MPILinearOperator`

rohanbabbar04 commented:
Hi @mrava87, 
- I have added `MPILinearOperator` class and made `MPIBlockDiag` extend that.
- Added important classes(`_ScaledLinearOperator`,` _AdjointLinearOperator`, etc) and made them input and output `DistributedArray`.
- For the linearop tests(for Scaled, Prod and Sum) I have used `MPIBlockDiag`, as our linearop.

mrava87 commented on tests/test_linearop.py:
```diff
@@ -0,0 +1,254 @@
+
+@pytest.mark.mpi(min_size=2)
+@pytest.mark.parametrize("par", [(par1), (par1j)])
+def test_square_linearop(par):
```
Add short doctoring explaining what this test is

rohanbabbar04 replied:
Done

mrava87 commented on tests/test_linearop.py:
```diff
@@ -0,0 +1,254 @@
+
+@pytest.mark.mpi(min_size=2)
+@pytest.mark.parametrize("par", [(par1), (par1j), (par2), (par2j)])
+def test_linearop(par):
```
Add short doctoring explaining what this test is

rohanbabbar04 replied:
Done

mrava87 reviewed (commented):
This looks great!

I made just a few comments on the LinearOperator. Most should be easy to handle, but there is one or two where I cannot understand some of your code choices that would be good to discuss/agree upon before moving ahead.

After that, for me this is ready to be merged :)

rohanbabbar04 commented on pylops_mpi/LinearOperator.py:
```diff
@@ -0,0 +1,332 @@
+from pylops_mpi import DistributedArray
+
+
+class MPILinearOperator(LinearOperator):
```
Made MPILinearOperator a standalone class

rohanbabbar04 commented on pylops_mpi/LinearOperator.py:
```diff
@@ -0,0 +1,332 @@
+
+    def __init__(self, shape: ShapeLike, dtype: DTypeLike, Op: LinearOperator = None,
+                 base_comm: MPI.Comm = MPI.COMM_WORLD):
+        self.Op = Op
```
Since I have made a standalone class
So keeping it as `self.Op = Op`

rohanbabbar04 commented on pylops_mpi/LinearOperator.py:
```diff
@@ -0,0 +1,332 @@
+                x = x.local_array
+            y[:] = self.Op._matvec(x)
+        else:
+            y = self._matvec(x)
```
Added `_matvec` and `_rmatvec` just like we do in pylops...

rohanbabbar04 commented:
- I have made `MPILinearOperator` class a standalone class(not extending `pylops.LinearOperato`r), have add some docstrings comment for different methods as a reason.
- Currently dot only supports matvec operations, not matmat.
- Thinking about `asmpilinearoperator`, currently reverted from `self.base_comm.allreduce(self.shape[1])` to `self.shape[1]`, but will have to handled for different operators as a direct self.Op._matvec(x.local_array) will possibly give an error...
- For the `MPILinearOperator`, there are mainly 3 parameters we primarily use which was in `pylops.LinearOperator`, which are Op, shape and dtype..

mrava87 commented:
Great, I think this is ready to go. We will see if things like `matmat` are needed in future but for now what we have in `MPILinearOperator` should be enough to be able to successfully solve our initial use-cases :)
